src/Moq/ExpressionExtensions.cs

```diff
@@ -17,6 +17,38 @@ namespace Moq
 {
 	internal static partial class ExpressionExtensions
 	{
+		/// <summary>
+		///   Wraps this <paramref name="expression"/> in a <see cref="ExpressionType.Convert"/> node if needed.
+		/// </summary>
+		/// <param name="expression">The <see cref="Expression"/> which should be wrapped.</param>
+		/// <param name="type">The <see cref="Type"/> with which to make the <paramref name="expression"/> compatible.</param>
+		/// <remarks>
+		///   LINQ expression trees generally enforce type compatibility rules that are stricter than
+		///   the assignment-compatibility used by e.g. <see cref="Type.IsAssignableFrom(Type)"/>. For
+		///   example, while <see langword="int"/> is assignable-to <see langword="object"/>, you
+		///   will need a conversion in a LINQ expression tree to model the value-type boxing operation.
+		/// </remarks>
+		internal static Expression ConvertIfNeeded(this Expression expression, Type type)
```

Comment:
* Check whether this method can be reused in `ActionObserver`, which currently has [its own (but slightly different) logic](https://github.com/moq/moq4/blob/0aa95749a97a9504e0e3adc12b398c02301b0ec5/src/Moq/ActionObserver.cs#L186-L209) for doing the same thing.

 * Also, add a set of tests that test our assumptions (stated here informally inside `<remarks>`) about when `Convert` nodes are necessary.